From 33eafc7a9e84098c61eb49514b3039f50b9a1373 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 20 Jan 2022 15:19:01 +0200
Subject: [PATCH] [SW-69581] pci: use device struct for dev specific data

Including device specific headers from common code is not allowed
anymore, and instead the device structures should be used.
PCI device code is refactored to use device structure.

Change-Id: Ie5a0ffbe9adf2a41925ca2d42417733db79a5f86
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/app/ext_rom/erom_pcie.c b/app/ext_rom/erom_pcie.c
index 705c869..97f89eb 100644
--- a/app/ext_rom/erom_pcie.c
+++ b/app/ext_rom/erom_pcie.c
@@ -14,6 +14,7 @@
 static void device_init(void)
 {
 	device_init_uart();
+	device_init_pci();
 }
 
 void erom_main_function(void)
diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index db39f4f..0a3f89e 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -7,7 +7,6 @@
 
 #include <string.h>
 #include <psoc_global_conf_regs.h>
-#include <pcie_aux_regs.h>
 #include <cpu_ca53_cfg_regs.h>
 #include <hl_boot_if.h>
 #include <bitops.h>
@@ -169,8 +168,8 @@
 		return false;
 
 	prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
-	smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
-	rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
+	smlh = regmap_readl(pci.aux_rdlh_link_up_reg);
+	rdlh = regmap_readl(pci.aux_smlh_link_up_reg);
 
 	if (prstn && smlh && rdlh)
 		return true;
@@ -266,6 +265,7 @@
 {
 	device_init_efuse();
 	device_init_uart();
+	device_init_pci();
 }
 
 __weak void ppboot_security_init(void) {}
diff --git a/app/preboot/preboot.c b/app/preboot/preboot.c
index 3ba446b..a0249e8 100644
--- a/app/preboot/preboot.c
+++ b/app/preboot/preboot.c
@@ -413,6 +413,7 @@
 {
 	device_init_efuse();
 	device_init_uart();
+	device_init_pci();
 }
 
 __weak void hl_set_reset_cause(uint32_t reset_cause) {}
diff --git a/arch/gaudi/arm/common.c b/arch/gaudi/arm/common.c
index ada1a26..f498919 100644
--- a/arch/gaudi/arm/common.c
+++ b/arch/gaudi/arm/common.c
@@ -660,3 +660,21 @@
 	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
 	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
 }
+
+void device_init_pci(void)
+{
+	pci.dbi_bar0_reg = mmPCIE_DBI_BAR0_REG;
+	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
+	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
+	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
+	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
+	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
+	pci.aux_ltssm_en_reg = mmPCIE_AUX_LTSSM_EN;
+	pci.aux_pf_flr_active_reg = mmPCIE_AUX_PF_FLR_ACTIVE;
+	pci.aux_flr_int_reg = mmPCIE_AUX_FLR_INT;
+	pci.aux_pf_flr_done_reg = mmPCIE_AUX_PF_FLR_DONE;
+	pci.aux_rdlh_link_up_reg = mmPCIE_AUX_RDLH_LINK_UP;
+	pci.aux_smlh_link_up_reg = mmPCIE_AUX_SMLH_LINK_UP;
+	pci.core_hot_rst_reg = mmPCIE_CORE_HOT_RST;
+}
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index d89862c..aa550d7 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -1229,3 +1229,21 @@
 	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
 	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
 }
+
+void device_init_pci(void)
+{
+	pci.dbi_bar0_reg = mmPCIE_DBI_BAR0_REG;
+	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
+	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
+	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
+	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
+	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
+	pci.aux_ltssm_en_reg = mmPCIE_AUX_LTSSM_EN;
+	pci.aux_pf_flr_active_reg = mmPCIE_AUX_PF_FLR_ACTIVE;
+	pci.aux_flr_int_reg = mmPCIE_AUX_FLR_INT;
+	pci.aux_pf_flr_done_reg = mmPCIE_AUX_PF_FLR_DONE;
+	pci.aux_rdlh_link_up_reg = mmPCIE_AUX_RDLH_LINK_UP;
+	pci.aux_smlh_link_up_reg = mmPCIE_AUX_SMLH_LINK_UP;
+	pci.core_hot_rst_reg = mmPCIE_CORE_HOT_RST;
+}
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 6c4003e..cef430d 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -8,6 +8,8 @@
 #include <d0_psoc_gpio0_regs.h>
 #include <d0_psoc_qspi_regs.h>
 #include <d0_pcie_aux_regs.h>
+#include <d0_pcie_core_regs.h>
+#include <d0_pcie_dbi_sig_regs.h>
 #include <d0_psoc_reset_conf_regs.h>
 #include <d0_psoc_efuse_regs.h>
 #include <d0_psoc_uart_0_regs.h>
@@ -154,6 +156,24 @@
 	uart.uart_bauddiv_reg = mmD0_PSOC_UART_0_BAUDDIV;
 }
 
+void device_init_pci(void)
+{
+	pci.dbi_bar0_reg = mmD0_PCIE_DBI_SIG_BAR0_REG;
+	pci.dbi_bar2_reg = mmD0_PCIE_DBI_SIG_BAR2_REG;
+	pci.dbi_bar4_reg = mmD0_PCIE_DBI_SIG_BAR4_REG;
+	pci.dbi_status_command_reg = mmD0_PCIE_DBI_SIG_STATUS_COMMAND_REG;
+	pci.aux_dbi_reg = mmD0_PCIE_AUX_DBI;
+	pci.aux_mem_space_en_reg = mmD0_PCIE_AUX_MEM_SPACE_EN;
+	pci.aux_flr_ctrl_reg = mmD0_PCIE_AUX_FLR_CTRL;
+	pci.aux_ltssm_en_reg = mmD0_PCIE_AUX_LTSSM_EN;
+	pci.aux_pf_flr_active_reg = mmD0_PCIE_AUX_PF_FLR_ACTIVE;
+	pci.aux_flr_int_reg = mmD0_PCIE_AUX_FLR_INT;
+	pci.aux_pf_flr_done_reg = mmD0_PCIE_AUX_PF_FLR_DONE;
+	pci.aux_rdlh_link_up_reg = mmD0_PCIE_AUX_RDLH_LINK_UP;
+	pci.aux_smlh_link_up_reg = mmD0_PCIE_AUX_SMLH_LINK_UP;
+	pci.core_hot_rst_reg = mmD0_PCIE_CORE_HOT_RST;
+}
+
 void lkd_dev_specific_setup(void)
 {
 	cpu_boot_dev_sts_set(CPU_BOOT_DEV_STS_FW_HARD_RST_EN);
diff --git a/arch/goya/arm/common.c b/arch/goya/arm/common.c
index 2ece7d5..060fc19 100644
--- a/arch/goya/arm/common.c
+++ b/arch/goya/arm/common.c
@@ -14,6 +14,7 @@
 #include <psoc_spi_regs.h>
 #include <psoc_uart_0_regs.h>
 #include <pcie_aux_regs.h>
+#include <pcie_dbi_regs.h>
 #include <pcie_core_regs.h>
 #include <cpu_timestamp_regs.h>
 #include <pcie_db_rsv_regs.h>
@@ -318,3 +319,21 @@
 	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
 	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
 }
+
+void device_init_pci(void)
+{
+	pci.dbi_bar0_reg = mmPCIE_DBI_BAR0_REG;
+	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
+	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
+	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
+	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
+	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
+	pci.aux_ltssm_en_reg = mmPCIE_AUX_LTSSM_EN;
+	pci.aux_pf_flr_active_reg = mmPCIE_AUX_PF_FLR_ACTIVE;
+	pci.aux_flr_int_reg = mmPCIE_AUX_FLR_INT;
+	pci.aux_pf_flr_done_reg = mmPCIE_AUX_PF_FLR_DONE;
+	pci.aux_rdlh_link_up_reg = mmPCIE_AUX_RDLH_LINK_UP;
+	pci.aux_smlh_link_up_reg = mmPCIE_AUX_SMLH_LINK_UP;
+	pci.core_hot_rst_reg = mmPCIE_CORE_HOT_RST;
+}
diff --git a/arch/greco/common/common.c b/arch/greco/common/common.c
index 2fe7992..ecab288 100644
--- a/arch/greco/common/common.c
+++ b/arch/greco/common/common.c
@@ -598,3 +598,21 @@
 	uart.uart_ctrl_reg = mmPSOC_UART_0_CTRL;
 	uart.uart_bauddiv_reg = mmPSOC_UART_0_BAUDDIV;
 }
+
+void device_init_pci(void)
+{
+	pci.dbi_bar0_reg = mmPCIE_DBI_BAR0_REG;
+	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
+	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
+	pci.dbi_status_command_reg = mmPCIE_DBI_STATUS_COMMAND_REG;
+	pci.aux_dbi_reg = mmPCIE_AUX_DBI;
+	pci.aux_mem_space_en_reg = mmPCIE_AUX_MEM_SPACE_EN;
+	pci.aux_flr_ctrl_reg = mmPCIE_AUX_FLR_CTRL;
+	pci.aux_ltssm_en_reg = mmPCIE_AUX_LTSSM_EN;
+	pci.aux_pf_flr_active_reg = mmPCIE_AUX_PF_FLR_ACTIVE;
+	pci.aux_flr_int_reg = mmPCIE_AUX_FLR_INT;
+	pci.aux_pf_flr_done_reg = mmPCIE_AUX_PF_FLR_DONE;
+	pci.aux_rdlh_link_up_reg = mmPCIE_AUX_RDLH_LINK_UP;
+	pci.aux_smlh_link_up_reg = mmPCIE_AUX_SMLH_LINK_UP;
+	pci.core_hot_rst_reg = mmPCIE_CORE_HOT_RST;
+}
diff --git a/include/device.h b/include/device.h
index 9cf3b6c..d410703 100644
--- a/include/device.h
+++ b/include/device.h
@@ -20,8 +20,27 @@
 	u32 uart_bauddiv_reg;
 };
 
+struct device_pci {
+	u32 dbi_bar0_reg;
+	u32 dbi_bar2_reg;
+	u32 dbi_bar4_reg;
+	u32 dbi_status_command_reg;
+	u32 aux_dbi_reg;
+	u32 aux_mem_space_en_reg;
+	u32 aux_flr_ctrl_reg;
+	u32 aux_ltssm_en_reg;
+	u32 aux_pf_flr_active_reg;
+	u32 aux_flr_int_reg;
+	u32 aux_pf_flr_done_reg;
+	u32 aux_rdlh_link_up_reg;
+	u32 aux_smlh_link_up_reg;
+	u32 core_hot_rst_reg;
+};
+
 extern struct device_efuse efuse;
 extern struct device_uart uart;
+extern struct device_pci pci;
 
 void device_init_efuse(void);
 void device_init_uart(void);
+void device_init_pci(void);
diff --git a/lib/common/pci.c b/lib/common/pci.c
index bad97cc..a821987 100644
--- a/lib/common/pci.c
+++ b/lib/common/pci.c
@@ -5,12 +5,10 @@
  *
  */
 
-#include <pcie_dbi_regs.h>
-#include <pcie_aux_regs.h>
 #include <common/pci_ids.h>
-#include <pcie_dbi_masks.h>
 #include <bitops.h>
 #include <preboot.h>
+#include <device.h>
 #include <sizes.h>
 #include "pci.h"
 
@@ -37,6 +35,8 @@
 #define IATU_OUTBOUND_REGION_SIZE	SZ_8K
 #define IATU_OUTBOUND_UPPER_ADDR_VAL	0x08000000
 
+struct device_pci pci;
+
 __weak u32 dev_pci_dbi_cfg_offset_get(void)
 {
 	return 0;
@@ -59,25 +59,25 @@
 
 	switch (bar_id) {
 	case PCI_BAR_ID_0:
-		bar_reg = mmPCIE_DBI_BAR0_REG;
+		bar_reg = pci.dbi_bar0_reg;
 		break;
 	case PCI_BAR_ID_2:
-		bar_reg = mmPCIE_DBI_BAR2_REG;
+		bar_reg = pci.dbi_bar2_reg;
 		break;
 	case PCI_BAR_ID_4:
-		bar_reg = mmPCIE_DBI_BAR4_REG;
+		bar_reg = pci.dbi_bar4_reg;
 		break;
 	default:
 		return -1;
 	}
 
-	aux_dbi_reg = regmap_readl(mmPCIE_AUX_DBI);
-	regmap_writel(IATU_ACCESS_DISABLE_VAL, mmPCIE_AUX_DBI);
+	aux_dbi_reg = regmap_readl(pci.aux_dbi_reg);
+	regmap_writel(IATU_ACCESS_DISABLE_VAL, pci.aux_dbi_reg);
 
 	bar_addr_low = regmap_readl(bar_reg) & ~0xF;
 	bar_addr_high = regmap_readl(bar_reg + sizeof(u32));
 
-	regmap_writel(aux_dbi_reg, mmPCIE_AUX_DBI);
+	regmap_writel(aux_dbi_reg, pci.aux_dbi_reg);
 
 	*bar_phys_addr = (u64)bar_addr_high << 32 | bar_addr_low;
 
@@ -97,9 +97,9 @@
 	u64 bar_phys_base = 0, region_base, region_end_address;
 	u32 offset = dev_pci_dbi_cfg_offset_get();
 	u32 ctrl_reg_val;
-	u32 aux_dbi_reg = regmap_readl(mmPCIE_AUX_DBI);
+	u32 aux_dbi_reg = regmap_readl(pci.aux_dbi_reg);
 
-	regmap_writel(IATU_ACCESS_ENABLE_VAL, mmPCIE_AUX_DBI);
+	regmap_writel(IATU_ACCESS_ENABLE_VAL, pci.aux_dbi_reg);
 
 	/* region offset */
 	offset += (IATU_REGION_SIZE * region) + IATU_REGION_INBOUND_OFF;
@@ -143,7 +143,7 @@
 	regmap_writel(ctrl_reg_val, offset + IATU_REGION_CTRL_2_OFF);
 
 	/* restore iATU access state */
-	regmap_writel(aux_dbi_reg, mmPCIE_AUX_DBI);
+	regmap_writel(aux_dbi_reg, pci.aux_dbi_reg);
 }
 
 /*
@@ -156,9 +156,9 @@
 {
 	u64 outbound_region_end_address;
 	u32 offset = dev_pci_dbi_cfg_offset_get();
-	u32 aux_dbi_reg = regmap_readl(mmPCIE_AUX_DBI);
+	u32 aux_dbi_reg = regmap_readl(pci.aux_dbi_reg);
 
-	regmap_writel(IATU_ACCESS_ENABLE_VAL, mmPCIE_AUX_DBI);
+	regmap_writel(IATU_ACCESS_ENABLE_VAL, pci.aux_dbi_reg);
 
 	/* disable region before configuring */
 	regmap_writel(0, offset + IATU_REGION_CTRL_2_OFF);
@@ -183,7 +183,7 @@
 	regmap_writel(IATU_REGION_EN_MASK, offset + IATU_REGION_CTRL_2_OFF);
 
 	/* restore iATU access state */
-	regmap_writel(aux_dbi_reg, mmPCIE_AUX_DBI);
+	regmap_writel(aux_dbi_reg, pci.aux_dbi_reg);
 }
 
 /*
@@ -191,7 +191,7 @@
  */
 __weak bool hl_pci_is_mem_enabled(void)
 {
-	u32 mem_space_ena_reg = regmap_readl(mmPCIE_AUX_MEM_SPACE_EN);
+	u32 mem_space_ena_reg = regmap_readl(pci.aux_mem_space_en_reg);
 
 	return !!mem_space_ena_reg;
 }
@@ -203,14 +203,16 @@
 	return (dev_id << 16) | PCI_VENDOR_ID_HABANALABS;
 }
 
+#define PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT	1
+
 void hl_pcie_disable_mem_access(void)
 {
 	u32 reg;
 
 	/* Disable the PCIe inbound memory requests */
-	reg = regmap_readl(mmPCIE_DBI_STATUS_COMMAND_REG);
+	reg = regmap_readl(pci.dbi_status_command_reg);
 	reg &= ~BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
-	regmap_writel(reg, mmPCIE_DBI_STATUS_COMMAND_REG);
+	regmap_writel(reg, pci.dbi_status_command_reg);
 }
 
 void hl_pcie_enable_mem_access(void)
@@ -218,7 +220,7 @@
 	u32 reg;
 
 	/* Enable the PCIe inbound memory requests */
-	reg = regmap_readl(mmPCIE_DBI_STATUS_COMMAND_REG);
+	reg = regmap_readl(pci.dbi_status_command_reg);
 	reg |= BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
-	regmap_writel(reg, mmPCIE_DBI_STATUS_COMMAND_REG);
+	regmap_writel(reg, pci.dbi_status_command_reg);
 }
diff --git a/lib/preboot/module_pci.c b/lib/preboot/module_pci.c
index 1e5b3c2..450653f 100644
--- a/lib/preboot/module_pci.c
+++ b/lib/preboot/module_pci.c
@@ -4,12 +4,10 @@
  * All Rights Reserved.
  */
 
-#include <pcie_aux_regs.h>
-#include <pcie_aux_masks.h>
-#include <pcie_core_regs.h>
 
 #include <bitops.h>
 #include <preboot.h>
+#include <device.h>
 
 #include "cmd_module.h"
 #include "uart.h"
@@ -43,25 +41,28 @@
 __weak void pci_dev_specific_disable(void) {}
 __weak void pci_dev_specific_enable(void) {}
 
+#define PCIE_AUX_FLR_CTRL_HW_CTRL_MASK	0x1
+#define PCIE_AUX_FLR_CTRL_INT_MASK_MASK	0x2
+
 __weak void pci_dev_flr_enable(bool enable)
 {
 	u32 val = enable ? 0 : PCIE_AUX_FLR_CTRL_HW_CTRL_MASK |
 				PCIE_AUX_FLR_CTRL_INT_MASK_MASK;
 
-	regmap_writel(val, mmPCIE_AUX_FLR_CTRL);
+	regmap_writel(val, pci.aux_flr_ctrl_reg);
 }
 
 __weak void pci_dev_specific_hot_rst(void)
 {
-	regmap_writel(0x10, mmPCIE_CORE_HOT_RST);
-	regmap_writel(1, mmPCIE_AUX_LTSSM_EN);
+	regmap_writel(0x10, pci.core_hot_rst_reg);
+	regmap_writel(1, pci.aux_ltssm_en_reg);
 }
 
 __weak void pci_dev_specific_flr(void)
 {
-	regmap_writel(1, mmPCIE_AUX_PF_FLR_ACTIVE);
-	regmap_writel(1, mmPCIE_AUX_FLR_INT);
-	regmap_writel(1, mmPCIE_AUX_PF_FLR_DONE);
+	regmap_writel(1, pci.aux_pf_flr_active_reg);
+	regmap_writel(1, pci.aux_flr_int_reg);
+	regmap_writel(1, pci.aux_pf_flr_done_reg);
 }
 
 __weak void hl_pci_iatu_init(void) {}
@@ -98,7 +99,7 @@
 
 __weak u32 pci_dev_specific_hot_rst_check(void)
 {
-	return regmap_readl(mmPCIE_CORE_HOT_RST) & BIT(0);
+	return regmap_readl(pci.core_hot_rst_reg) & BIT(0);
 }
 
 __weak enum cmd_check_res dev_specific_module_pci_check(void)
@@ -112,10 +113,10 @@
 		return NOT_FOUND;
 	}
 
-	pci_sts.flr_reg_val = regmap_readl(mmPCIE_AUX_FLR_INT);
+	pci_sts.flr_reg_val = regmap_readl(pci.aux_flr_int_reg);
 	pci_sts.hot_rst_reg_val = pci_dev_specific_hot_rst_check();
-	pci_sts.rdlh_link_reg_val = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
-	pci_sts.smlh_link_reg_val = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+	pci_sts.rdlh_link_reg_val = regmap_readl(pci.aux_rdlh_link_up_reg);
+	pci_sts.smlh_link_reg_val = regmap_readl(pci.aux_smlh_link_up_reg);
 
 	if (!pci_sts.pci_pll_initialized)
 		return FOUND_SLOW_RETURN;
